### PR TITLE
Adds PDF files to Modular Computers

### PR DESCRIPTION
**Pull request by Fruerlund**
You can now scan files into modular computers.
That means papers that looks very nice such as detective reports, body scanners can be scanned into the modular computer and sent, printed or whatever. What you cannot do is edit them, if you try to you will be warned.

Computers/tablets bought at the computer vendor will now include a portable USB device.

🆑Fruerlund
feature: You can now scan papers into modular computers.
feature: Laptops and tablets now come with a portable device included.
/🆑

### CODE REVIEW

File: code/modules/modular_computers/hardware/printer.dm [hunk 1]
```diff
@@ -24,22 +27,34 @@
 	if(!check_functionality())
 		return FALSE
 
-	var/obj/item/weapon/paper/P = new/obj/item/weapon/paper(get_turf(holder))
+	if(!printing)
+		printing = 1
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: `TRUE`

File: code/modules/modular_computers/hardware/printer.dm [hunk 2]
```diff
@@ -24,22 +27,34 @@
...
+			P.info = stars(text_to_print, 100-malfunction_probability)
+		else
+			P.info = text_to_print
+		if(paper_title)
+			P.name = paper_title
+
+		P.update_icon()
+		stored_paper--
+		P = null
+		printing = 0
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: `FALSE`

File: code/modules/modular_computers/hardware/printer.dm [hunk 3]
```diff
@@ -24,22 +27,34 @@
...
-		P.info = stars(text_to_print, 100-malfunction_probability)
 	else
-		P.info = text_to_print
-	if(paper_title)
-		P.name = paper_title
-	P.update_icon()
-	stored_paper--
-	P = null
-	return TRUE
+		to_chat(usr, "<span class='notice'>[src] is already printing. Please wait</span>")
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: Missing a fullstop.

File: code/modules/modular_computers/hardware/printer.dm [hunk 4]
```diff
@@ -49,15 +64,31 @@
 			return FALSE
 
 		if(user)
-			to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
+			if(P.info == "")
+				to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
+			else
+				switch(alert("How would you like to proceed?", "Options.", "Scan Paper","Recycle Paper", "Abort"))
+					if("Abort")
+						return 0
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: `FALSE`

File: code/modules/modular_computers/hardware/printer.dm [hunk 5]
```diff
@@ -49,15 +64,31 @@
...
+					if("Scan Paper")
+						to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper scanner.</span>")
+						var/datum/computer_file/data/F = new/datum/computer_file/data()
+						F.filename = "[P.name] [worldtime2text()]"
+						F.filetype = "PDF"
+						F.stored_data = P.info
+						F.do_not_edit = 1
+						HDD.store_file(F)
+						return 0
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: `FALSE`

File: code/modules/modular_computers/hardware/printer.dm [hunk 6]
```diff
@@ -49,15 +64,31 @@
...
+				switch(alert("How would you like to proceed?", "Options.", "Scan Paper","Recycle Paper", "Abort"))
+					if("Abort")
+						return 0
+					if("Scan Paper")
+						to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper scanner.</span>")
+						var/datum/computer_file/data/F = new/datum/computer_file/data()
+						F.filename = "[P.name] [worldtime2text()]"
+						F.filetype = "PDF"
+						F.stored_data = P.info
+						F.do_not_edit = 1
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: boolean? `TRUE`

File: code/modules/modular_computers/laptop_vendor.dm [hunk 7]
```diff
@@ -92,6 +96,19 @@
 				if(fabricate)
 					fabricated_laptop.install_component(new /obj/item/weapon/computer_hardware/network_card/advanced)
 				total_price += 299
+		switch(dev_portable)
+			if(1) // Basic(16QC)
```
Comment (SamHPurp) [inline code/modules/modular_computers/laptop_vendor.dm]: I'd much rather these be defines than magic numbers... but thats the case for all of these vars :(

File: code/modules/modular_computers/laptop_vendor.dm [hunk 8]
```diff
@@ -201,6 +233,10 @@
 			dev_apc_recharger = text2num(href_list["tesla"])
 			fabricate_and_recalc_price(0)
 			return 1
+		if("hw_portable")
+			dev_portable = text2num(href_list["portable"])
+			fabricate_and_recalc_price(0)
+			return 1
```
Comment (SamHPurp) [inline code/modules/modular_computers/laptop_vendor.dm]: `TRUE`

File: code/modules/modular_computers/hardware/printer.dm [hunk 9]
```diff
@@ -49,15 +64,31 @@
...
+						HDD.store_file(F)
+						return 0
+					if("Recycle Paper")
+						to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
 		qdel(I)
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: Not part of your code... but is this gc'ing properly?

File: code/modules/modular_computers/hardware/printer.dm [hunk 10]
```diff
@@ -8,6 +8,9 @@
 	device_type = MC_PRINT
 	var/stored_paper = 20
 	var/max_paper = 30
+	var/HDRIVE = MC_HDD
+	var/printing = null										// Prevent printing spam
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: isn't this just boolean? couldn't/shouldn't this just be `FALSE`?

File: code/modules/modular_computers/hardware/printer.dm [hunk 11]
```diff
@@ -24,27 +27,32 @@
...
-		P.info = text_to_print
-	if(paper_title)
-		P.name = paper_title
-	P.update_icon()
-	P.populatefields()
-	P.updateinfolinks()
-	stored_paper--
-	P = null
+	if(!printing)
+		printing = 1
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: `TRUE`

File: code/modules/modular_computers/hardware/printer.dm [hunk 12]
```diff
@@ -24,27 +27,32 @@
...
+		else
+			P.info = text_to_print
+		if(paper_title)
+			P.name = paper_title
 
-	return TRUE
+		P.update_icon()
+		stored_paper--
+		P = null
+		printing = 0
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: `FALSE`

File: code/modules/modular_computers/hardware/printer.dm [hunk 13]
```diff
@@ -24,27 +27,32 @@
...
+		return TRUE
+	else
+		to_chat(usr, "<span class='notice'>[src] is already printing. Please wait</span>")
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: Missing fullstop

File: code/modules/modular_computers/hardware/printer.dm [hunk 14]
```diff
@@ -54,15 +62,32 @@
...
-			to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
+			if(P.info == null)
+				to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
+			else
+				user.put_in_hands(P)
+				switch(alert("How would you like to proceed?", "Options.", "Scan Paper","Recycle Paper", "Abort"))
+					if("Abort")
+						return FALSE
+					if("Scan Paper")
+						to_chat(user, "<span class='notice'>You scan \ [I] with [src]'s paper scanner.</span>")
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: should the `\` be `\the`?

File: code/modules/modular_computers/hardware/printer.dm [hunk 15]
```diff
@@ -24,40 +25,74 @@
...
+						to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
 		qdel(I)
 		stored_paper++
 		return TRUE
+	if(istype(I, /obj/item/weapon/paper_bundle))
+		var/obj/item/weapon/paper_bundle/B = I
+		var/datum/computer_file/data/F = new/datum/computer_file/data()
+		F.filename = "paper bundle - [B.name] [worldtime2text()]"
+		F.stored_data = ""
+		F.do_not_edit = 1
```
Comment (SamHPurp) [inline code/modules/modular_computers/hardware/printer.dm]: `TRUE`

File: code/modules/modular_computers/laptop_vendor.dm (hunk 8 above)
Comment (FalseIncarnate) [reply code/modules/modular_computers/laptop_vendor.dm]: return 1 is actually consistent with the rest of this proc, which either returns 0 or 1 based on success.

TONS of procs do this, such as attack and such. I'd say he's probably fine leaving this as a return 1 for consistency with not just the rest of the proc, but the rest of the codebase until we decided whether or not we want to make return TRUE or return FALSE a standardized styling thing

File: code/modules/modular_computers/hardware/printer.dm [hunk 16]
```diff
@@ -24,40 +25,74 @@
...
-		P.name = paper_title
-	P.update_icon()
-	P.populatefields()
-	P.updateinfolinks()
-	stored_paper--
-	P = null
 
-	playsound(loc, 'sound/goonstation/machines/printer_thermal.ogg', 50, 1)
+		if(damage > damage_malfunction)
+			P.info = stars(text_to_print, 100-malfunction_probability)
```
Comment (tigercat2000) [inline code/modules/modular_computers/hardware/printer.dm]: Spacing on operators please, `100 - malfunction_probability`

File: code/modules/modular_computers/hardware/printer.dm (hunk 4 above)
Comment (tigercat2000) [reply code/modules/modular_computers/hardware/printer.dm]: ^ FALSE

File: code/modules/modular_computers/hardware/printer.dm [hunk 17]
```diff
@@ -24,40 +25,74 @@
...
 			return FALSE
-
 		if(user && !user.unEquip(I))
 			return FALSE
-
 		if(user)
-			to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
+			user.put_in_hands(P)
+			if(P.info == null)
+				to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
```
Comment (tigercat2000) [inline code/modules/modular_computers/hardware/printer.dm]: `\the` is unnecessary

File: code/modules/modular_computers/hardware/printer.dm [hunk 18]
```diff
@@ -24,40 +25,74 @@
...
+			else
+				switch(alert("How would you like to proceed?", "Options.", "Scan Paper","Recycle Paper", "Abort"))
+					if("Abort")
+						return 0
+					if("Scan Paper")
+						to_chat(user, "<span class='notice'>You scan \the [I] with [src]'s paper scanner.</span>")
```
Comment (tigercat2000) [inline code/modules/modular_computers/hardware/printer.dm]: `\the` is unnecessary

File: code/modules/modular_computers/hardware/printer.dm [hunk 19]
```diff
@@ -24,40 +25,74 @@
...
+						var/datum/computer_file/data/F = new/datum/computer_file/data()
+						F.filename = "[P.name] [worldtime2text()]"
+						F.filetype = "PDF"
+						F.stored_data = P.info
+						F.do_not_edit = 1
+						HDD.store_file(F)
+						return 0
+					if("Recycle Paper")
+						to_chat(user, "<span class='notice'>You insert \the [I] into [src]'s paper recycler.</span>")
```
Comment (tigercat2000) [inline code/modules/modular_computers/hardware/printer.dm]: `\the` is unnecessary

**Comment by Fox-McCloud** [comment]
Been months and this still isn't updated; closing.

**Pull request closed without merging**